File: LinkedList/Middle of the Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 /*
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        float len = 0;
        ListNode* node = head;
        while(node->next != nullptr){
            node = node->next;
            len++;
        }
        int mid = ceil(len/2);
        node = head;
        for(int i = 0; i < mid; i++){
            node = node->next;
        }
        return node;
    }
};
*/
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode *fast = head, *slow = head;
        while(fast!=nullptr && fast->next!=nullptr){
            fast = fast->next->next;
            slow = slow->next;
        }
        return slow;
    }
};

File: LinkedList/Middle of the Linked List.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut count = 0_u32;
        let mut fastptr = &head;
        let mut slowptr = &head;
        while fastptr.is_some() {
            count += 1;
            if count % 2 == 0 {
                slowptr = &slowptr.as_ref().unwrap().next;
            }
            fastptr = &fastptr.as_ref().unwrap().next;
        }
        slowptr.clone()
    }
}